src/core/thread/meshcop_dataset_manager.cpp

```diff
@@ -382,13 +411,39 @@ ThreadError DatasetManager::Set(Coap::Header &aHeader, Message &aMessage, const
 
         aMessage.Read(offset, sizeof(Tlv), &data.tlv);
         aMessage.Read(offset + sizeof(Tlv), data.tlv.GetLength(), data.value);
-        mLocal.Set(data.tlv);
+
+        if (!isUpdateFromCommissioner && isTlvsAffectConnectivity)
+        {
+            mNetif.GetPendingDataset().GetNetwork().Set(data.tlv);;
+        }
+        else
+        {
+            mLocal.Set(data.tlv);
+        }
+
         offset += sizeof(Tlv) + data.tlv.GetLength();
     }
 
-    mNetwork = mLocal;
-    mNetworkDataLeader.IncrementVersion();
-    mNetworkDataLeader.IncrementStableVersion();
+    if (!isUpdateFromCommissioner && isTlvsAffectConnectivity)
+    {
+        DelayTimerTlv delaytimer;
+        uint8_t flags;
+
+        delaytimer.Init();
```

Comment:
The main reason that does not use Set directly is that we also need to update pending timestamp and delay timer. Till my understanding, Set would not copy TLVs from message and update them?